cartography/intel/gcp/crm.py

```diff
@@ -136,41 +137,67 @@ def load_gcp_projects(neo4j_session, data, gcp_update_tag):
-            elif project['parent']['type'] == "folder":
-                query = """
-                MERGE (parent:GCPFolder{id:{ParentId}})
-                ON CREATE SET parent.firstseen = timestamp()
-                """
-                parentid = f"folders/{project['parent']['id']}"
-        query += """
-        MERGE (project:GCPProject{id:{ProjectId}})
-        ON CREATE SET project.firstseen = timestamp()
-        SET project.projectid = {ProjectId},
-        project.projectnumber = {ProjectNumber},
-        project.displayname = {DisplayName},
-        project.lifecyclestate = {LifecycleState},
-        project.lastupdated = {gcp_update_tag}
-        WITH parent, project
-        MERGE (parent)-[r:RESOURCE]->(project)
-        ON CREATE SET r.firstseen = timestamp()
-        SET r.lastupdated = {gcp_update_tag}
-        """
-        neo4j_session.run(
-            query,
-            ParentId=parentid,
-            ProjectId=project['projectId'],
-            ProjectNumber=project['projectNumber'],
-            DisplayName=project.get('name', None),
-            LifecycleState=project.get('lifecycleState', None),
-            gcp_update_tag=gcp_update_tag,
-        )
+        if project.get('parent'):
+            # Project has parents, so set the parent node label to either `GCPOrganization` or `GCPFolder`.
+            if project['parent']['type'] == 'organization':
+                parent_label = 'GCPOrganization'
+            elif project['parent']['type'] == 'folder':
+                parent_label = 'GCPFolder'
+            else:
+                raise NotImplementedError(
+                    "Ingestion of GCP {}s as parent nodes is currently not supported. "
+                    "Please file an issue at https://github.com/lyft/cartography/issues.".format(
+                        project['parent']['type'],
+                    ),
+                )
+            # Create a parent node ID: either `organizations/{id}` or `folders/{id}`.
+            parentid = f"{project['parent']['type']}s/{project['parent']['id']}"
+
+            # Create parent node, create project node, and connect them together.
+            INGEST_PARENT_TEMPLATE = Template("""
+            MERGE (parent:$parent_label{id:{ParentId}})
+            ON CREATE SET parent.firstseen = timestamp()
+
+            MERGE (project:GCPProject{id:{ProjectId}})
+            ON CREATE SET project.firstseen = timestamp()
+            SET project.projectid = {ProjectId},
+            project.projectnumber = {ProjectNumber},
+            project.displayname = {DisplayName},
+            project.lifecyclestate = {LifecycleState},
+            project.lastupdated = {gcp_update_tag}
+
+            MERGE (parent)-[r:RESOURCE]->(project)
+            ON CREATE SET r.firstseen = timestamp()
+            SET r.lastupdated = {gcp_update_tag}
+            """)
+            neo4j_session.run(
+                INGEST_PARENT_TEMPLATE.safe_substitute(parent_label=parent_label),
```

Comment:
By the way, how was the neo4j driver handling the error from the query failing before? I feel like it might be safer to use `substitute()`, rather than `safe_substitute()`, since `substitute()` throws a `KeyError`, but I can't investigate right now how the driver deals with floating $s.

Reply:
Before, the error didn't even get to Neo4j because it failed at the python level:
![Screen Shot 2020-01-09 at 2 37 44 PM](https://user-images.githubusercontent.com/46503781/72178298-81dc5400-3397-11ea-93d1-86036e95c729.png)

Reply:
I don't recall seeing any errors in my console before, though. Maybe I just didn't set to `--verbose`?

But @marco-lancini, you likely woudn't encounter this bug if you are already using the GCP [organization resource](https://cloud.google.com/resource-manager/docs/creating-managing-organization), as all new projects would automatically go under the `organization`. But even if you've transitioned to using an `organization`, it's still possible for older projects to not belong, and be parentless.

Reply:
Yeah that's odd - if your project data didn't have a `parent` field, then the old code would have crashed with that UnboundLocalError because `query` wouldn't have been defined.

Reply:
This was back when I was working on #189, and I did my `EMPTYORG` workaround. But I'm glad you're doing this now.